toxcore/friend_requests.h

```diff
@@ -24,13 +24,39 @@
 #ifndef FRIEND_REQUESTS_H
 #define FRIEND_REQUESTS_H
 
+#include <stdbool.h>
 #include "friend_connection.h"
 
 #define MAX_FRIEND_REQUEST_DATA_SIZE (ONION_CLIENT_MAX_DATA_SIZE - (1 + sizeof(uint32_t)))
 
-typedef struct {
+/* Maximum nospam amount should be quite small to be effective against spam.
+ * Maximum allowed amount is given by MAX_NOSPAM_AMOUNT.
+ */
+
+#define MAX_NOSPAM_AMOUNT 16
+#define MAX_NOSPAM_DESCRIPTION_LENGTH 128
+#define NOSPAM_SIZE sizeof(uint32_t)
+#define NOSPAM_SPAM 0
+
+typedef enum NSERR {
+    NSERR_SUCCESS = 0,
+    NSERR_NOT_FOUND = 1,
+    NSERR_TOO_MANY = 2,
+    NSERR_ALREADY_EXISTS = 3,
+    NSERR_DESCRIPTION_TOO_LONG
+} NSERR;
+
+typedef struct No_Spam {
     uint32_t nospam;
-    void (*handle_friendrequest)(void *, const uint8_t *, const uint8_t *, size_t, void *);
+    uint8_t  description[MAX_NOSPAM_DESCRIPTION_LENGTH];
+    size_t description_length;
+} No_Spam;
+
+typedef struct {
+    No_Spam nospam[MAX_NOSPAM_AMOUNT];
```

Comment:
9 Why fixing the number of nospams to 16? Why didn't you use [list](https://github.com/irungentoo/toxcore/blob/master/toxcore/list.h) implementation of a resizable array already existing in toxcore to allow arbitrary number of nospams, possibly adding save and load methods for it?